TO DO : separate class wise and take equal number of classes or combine 3 and 4 and test on the remaining. 

Deal with the class Imbalance problem

In [2]:
from google.colab import files
uploaded = files.upload()

Saving winequality-white.csv to winequality-white.csv


In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import linear_model
import seaborn as sns
from sklearn import  metrics
from sklearn import preprocessing

In [0]:
white_data = pd.read_csv('winequality-white.csv', sep = ",")

In [0]:
white_data['quality'] = white_data['quality'].astype(object)

In [0]:
white_data['quality'] = white_data['quality'].astype(int)
white_data = pd.get_dummies(white_data, columns=['quality'])

In [0]:
train_data = white_data.iloc[:,0:11].values # first columns
train_labels = white_data.iloc[:,12:].values # last columns

In [0]:
train_data = preprocessing.normalize(train_data, axis=0)

In [0]:
train_data, test_data, train_value, test_value = train_test_split(train_data, train_labels, test_size = 0.33, random_state=42)

In [77]:
print(train_data.shape, test_data.shape, train_value.shape, test_value.shape)

(3281, 11) (1617, 11) (3281, 6) (1617, 6)


In [0]:
from sklearn.decomposition import PCA
pca = PCA(n_components=10)
data = pca.fit_transform(train_data)

In [0]:
test_data = pca.fit_transform(test_data)

In [0]:

from keras import layers, optimizers, regularizers
from keras.layers import Dense, Dropout, BatchNormalization, Activation
from keras.models import Sequential
# fix random seed for reproducibility
np.random.seed(7)

In [0]:
winemod1 = Sequential()
# layer 1
winemod1.add(Dense(32, input_dim=11, activation='relu', name='fc0',kernel_regularizer=regularizers.l2(0.01)))
winemod1.add(BatchNormalization(momentum=0.99, epsilon=0.001))
#layer 2
winemod1.add(Dense(64, name='fc1',bias_initializer='zeros'))
winemod1.add(BatchNormalization(momentum=0.99, epsilon=0.001))
winemod1.add(Activation('tanh'))
winemod1.add(Dropout(0.3))
#layer 3
winemod1.add(Dense(128, name='fc2',bias_initializer='zeros'))
winemod1.add(BatchNormalization(momentum=0.99, epsilon=0.001))
winemod1.add(Activation('relu'))
winemod1.add(Dropout(0.3))
#layer 4
winemod1.add(Dense(6, name='fc3',bias_initializer='zeros'))
winemod1.add(BatchNormalization(momentum=0.99, epsilon=0.001))
winemod1.add(Activation('softmax'))


In [90]:
winemod1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
fc0 (Dense)                  (None, 32)                384       
_________________________________________________________________
batch_normalization_25 (Batc (None, 32)                128       
_________________________________________________________________
fc1 (Dense)                  (None, 64)                2112      
_________________________________________________________________
batch_normalization_26 (Batc (None, 64)                256       
_________________________________________________________________
activation_19 (Activation)   (None, 64)                0         
_________________________________________________________________
dropout_13 (Dropout)         (None, 64)                0         
_________________________________________________________________
fc2 (Dense)                  (None, 128)               8320      
__________

In [0]:
Adam = optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0)
winemod1.compile(optimizer = Adam, loss = "categorical_crossentropy", metrics = ["categorical_accuracy"])

In [93]:
# Fit the model
winemod1.fit(train_data, train_value, epochs=200, batch_size=64, validation_data=(test_data, test_value))

Train on 3281 samples, validate on 1617 samples
Epoch 1/200
3281/3281 [==============================] - 2s 617us/step - loss: 2.0416 - categorical_accuracy: 0.2490 - val_loss: 2.2129 - val_categorical_accuracy: 0.1694
Epoch 2/200
3281/3281 [==============================] - 0s 76us/step - loss: 1.7601 - categorical_accuracy: 0.3737 - val_loss: 1.7634 - val_categorical_accuracy: 0.4162
Epoch 3/200
3281/3281 [==============================] - 0s 71us/step - loss: 1.6703 - categorical_accuracy: 0.4230 - val_loss: 1.6198 - val_categorical_accuracy: 0.4774
Epoch 4/200
3281/3281 [==============================] - 0s 74us/step - loss: 1.5947 - categorical_accuracy: 0.4547 - val_loss: 1.5622 - val_categorical_accuracy: 0.5207
Epoch 5/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.5436 - categorical_accuracy: 0.4752 - val_loss: 1.4920 - val_categorical_accuracy: 0.4719
Epoch 6/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.4898 - categorical_ac

3281/3281 [==============================] - 0s 73us/step - loss: 1.2615 - categorical_accuracy: 0.5206 - val_loss: 1.3260 - val_categorical_accuracy: 0.4824
Epoch 16/200
3281/3281 [==============================] - 0s 74us/step - loss: 1.2410 - categorical_accuracy: 0.5306 - val_loss: 1.2904 - val_categorical_accuracy: 0.4886
Epoch 17/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.2354 - categorical_accuracy: 0.5340 - val_loss: 1.2477 - val_categorical_accuracy: 0.5003
Epoch 18/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.2203 - categorical_accuracy: 0.5404 - val_loss: 1.2147 - val_categorical_accuracy: 0.5250
Epoch 19/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.2054 - categorical_accuracy: 0.5419 - val_loss: 1.2882 - val_categorical_accuracy: 0.4793
Epoch 20/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.2000 - categorical_accuracy: 0.5288 - val_loss: 1.1650 - val_categorical_accu

Epoch 30/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.1133 - categorical_accuracy: 0.5608 - val_loss: 1.1269 - val_categorical_accuracy: 0.5232
Epoch 31/200
3281/3281 [==============================] - 0s 76us/step - loss: 1.1123 - categorical_accuracy: 0.5547 - val_loss: 1.1485 - val_categorical_accuracy: 0.5281
Epoch 32/200
3281/3281 [==============================] - 0s 71us/step - loss: 1.1081 - categorical_accuracy: 0.5511 - val_loss: 1.2045 - val_categorical_accuracy: 0.4904
Epoch 33/200
3281/3281 [==============================] - 0s 74us/step - loss: 1.1031 - categorical_accuracy: 0.5401 - val_loss: 1.0857 - val_categorical_accuracy: 0.5405
Epoch 34/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.1016 - categorical_accuracy: 0.5443 - val_loss: 1.2479 - val_categorical_accuracy: 0.4601
Epoch 35/200
3281/3281 [==============================] - 0s 76us/step - loss: 1.0919 - categorical_accuracy: 0.5514 - val_loss: 1.0955 - val_cat

3281/3281 [==============================] - 0s 71us/step - loss: 1.0702 - categorical_accuracy: 0.5584 - val_loss: 1.1152 - val_categorical_accuracy: 0.5028
Epoch 45/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0660 - categorical_accuracy: 0.5623 - val_loss: 1.1019 - val_categorical_accuracy: 0.5071
Epoch 46/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0724 - categorical_accuracy: 0.5553 - val_loss: 1.0843 - val_categorical_accuracy: 0.5145
Epoch 47/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0559 - categorical_accuracy: 0.5596 - val_loss: 1.0855 - val_categorical_accuracy: 0.5108
Epoch 48/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0625 - categorical_accuracy: 0.5547 - val_loss: 1.1047 - val_categorical_accuracy: 0.5201
Epoch 49/200
3281/3281 [==============================] - 0s 70us/step - loss: 1.0628 - categorical_accuracy: 0.5669 - val_loss: 1.0944 - val_categorical_accu

Epoch 59/200
3281/3281 [==============================] - 0s 77us/step - loss: 1.0489 - categorical_accuracy: 0.5575 - val_loss: 1.0464 - val_categorical_accuracy: 0.5318
Epoch 60/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0362 - categorical_accuracy: 0.5672 - val_loss: 1.0469 - val_categorical_accuracy: 0.5343
Epoch 61/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0417 - categorical_accuracy: 0.5562 - val_loss: 1.0994 - val_categorical_accuracy: 0.5442
Epoch 62/200
3281/3281 [==============================] - 0s 74us/step - loss: 1.0359 - categorical_accuracy: 0.5635 - val_loss: 1.0484 - val_categorical_accuracy: 0.5448
Epoch 63/200
3281/3281 [==============================] - 0s 71us/step - loss: 1.0350 - categorical_accuracy: 0.5620 - val_loss: 1.0491 - val_categorical_accuracy: 0.5294
Epoch 64/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0372 - categorical_accuracy: 0.5584 - val_loss: 1.0493 - val_cat

3281/3281 [==============================] - 0s 74us/step - loss: 1.0326 - categorical_accuracy: 0.5648 - val_loss: 1.0555 - val_categorical_accuracy: 0.5164
Epoch 74/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0333 - categorical_accuracy: 0.5666 - val_loss: 1.0383 - val_categorical_accuracy: 0.5399
Epoch 75/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0333 - categorical_accuracy: 0.5553 - val_loss: 1.0652 - val_categorical_accuracy: 0.5102
Epoch 76/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0280 - categorical_accuracy: 0.5608 - val_loss: 1.0384 - val_categorical_accuracy: 0.5325
Epoch 77/200
3281/3281 [==============================] - 0s 77us/step - loss: 1.0218 - categorical_accuracy: 0.5678 - val_loss: 1.0771 - val_categorical_accuracy: 0.5133
Epoch 78/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0308 - categorical_accuracy: 0.5599 - val_loss: 1.0351 - val_categorical_accu

3281/3281 [==============================] - 0s 75us/step - loss: 1.0233 - categorical_accuracy: 0.5590 - val_loss: 1.0468 - val_categorical_accuracy: 0.5529
Epoch 88/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0253 - categorical_accuracy: 0.5611 - val_loss: 1.0370 - val_categorical_accuracy: 0.5430
Epoch 89/200
3281/3281 [==============================] - 0s 74us/step - loss: 1.0191 - categorical_accuracy: 0.5626 - val_loss: 1.0487 - val_categorical_accuracy: 0.5442
Epoch 90/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0150 - categorical_accuracy: 0.5642 - val_loss: 1.0356 - val_categorical_accuracy: 0.5411
Epoch 91/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0208 - categorical_accuracy: 0.5608 - val_loss: 1.0371 - val_categorical_accuracy: 0.5349
Epoch 92/200
3281/3281 [==============================] - 0s 77us/step - loss: 1.0192 - categorical_accuracy: 0.5687 - val_loss: 1.0348 - val_categorical_accu

3281/3281 [==============================] - 0s 73us/step - loss: 1.0063 - categorical_accuracy: 0.5684 - val_loss: 1.1685 - val_categorical_accuracy: 0.5015
Epoch 102/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0182 - categorical_accuracy: 0.5626 - val_loss: 1.0803 - val_categorical_accuracy: 0.5269
Epoch 103/200
3281/3281 [==============================] - 0s 76us/step - loss: 1.0115 - categorical_accuracy: 0.5596 - val_loss: 1.1349 - val_categorical_accuracy: 0.5114
Epoch 104/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.0190 - categorical_accuracy: 0.5632 - val_loss: 1.1084 - val_categorical_accuracy: 0.5294
Epoch 105/200
3281/3281 [==============================] - 0s 78us/step - loss: 1.0075 - categorical_accuracy: 0.5626 - val_loss: 1.1721 - val_categorical_accuracy: 0.4972
Epoch 106/200
3281/3281 [==============================] - 0s 82us/step - loss: 1.0139 - categorical_accuracy: 0.5690 - val_loss: 1.0648 - val_categorical

3281/3281 [==============================] - 0s 75us/step - loss: 1.0028 - categorical_accuracy: 0.5736 - val_loss: 1.0283 - val_categorical_accuracy: 0.5560
Epoch 116/200
3281/3281 [==============================] - 0s 72us/step - loss: 1.0002 - categorical_accuracy: 0.5724 - val_loss: 1.1072 - val_categorical_accuracy: 0.4886
Epoch 117/200
3281/3281 [==============================] - 0s 73us/step - loss: 1.0071 - categorical_accuracy: 0.5611 - val_loss: 1.0543 - val_categorical_accuracy: 0.5238
Epoch 118/200
3281/3281 [==============================] - 0s 72us/step - loss: 0.9916 - categorical_accuracy: 0.5785 - val_loss: 1.0668 - val_categorical_accuracy: 0.5195
Epoch 119/200
3281/3281 [==============================] - 0s 77us/step - loss: 1.0007 - categorical_accuracy: 0.5645 - val_loss: 1.0436 - val_categorical_accuracy: 0.5485
Epoch 120/200
3281/3281 [==============================] - 0s 75us/step - loss: 1.0053 - categorical_accuracy: 0.5709 - val_loss: 1.0366 - val_categorical

3281/3281 [==============================] - 0s 79us/step - loss: 0.9970 - categorical_accuracy: 0.5776 - val_loss: 1.0267 - val_categorical_accuracy: 0.5498
Epoch 130/200
3281/3281 [==============================] - 0s 76us/step - loss: 1.0046 - categorical_accuracy: 0.5626 - val_loss: 1.1801 - val_categorical_accuracy: 0.4898
Epoch 131/200
3281/3281 [==============================] - 0s 78us/step - loss: 0.9862 - categorical_accuracy: 0.5763 - val_loss: 1.0480 - val_categorical_accuracy: 0.5269
Epoch 132/200
3281/3281 [==============================] - 0s 80us/step - loss: 0.9924 - categorical_accuracy: 0.5718 - val_loss: 1.0558 - val_categorical_accuracy: 0.5399
Epoch 133/200
3281/3281 [==============================] - 0s 74us/step - loss: 0.9826 - categorical_accuracy: 0.5788 - val_loss: 1.2211 - val_categorical_accuracy: 0.4873
Epoch 134/200
3281/3281 [==============================] - 0s 73us/step - loss: 0.9925 - categorical_accuracy: 0.5767 - val_loss: 1.1174 - val_categorical

Epoch 144/200
3281/3281 [==============================] - 0s 69us/step - loss: 0.9973 - categorical_accuracy: 0.5672 - val_loss: 1.0336 - val_categorical_accuracy: 0.5479
Epoch 145/200
3281/3281 [==============================] - 0s 68us/step - loss: 0.9842 - categorical_accuracy: 0.5760 - val_loss: 1.0494 - val_categorical_accuracy: 0.5238
Epoch 146/200
3281/3281 [==============================] - 0s 69us/step - loss: 0.9819 - categorical_accuracy: 0.5770 - val_loss: 1.0408 - val_categorical_accuracy: 0.5207
Epoch 147/200
3281/3281 [==============================] - 0s 72us/step - loss: 0.9791 - categorical_accuracy: 0.5837 - val_loss: 1.0422 - val_categorical_accuracy: 0.5448
Epoch 148/200
3281/3281 [==============================] - 0s 80us/step - loss: 0.9846 - categorical_accuracy: 0.5821 - val_loss: 1.0496 - val_categorical_accuracy: 0.5380
Epoch 149/200
3281/3281 [==============================] - 0s 75us/step - loss: 0.9828 - categorical_accuracy: 0.5742 - val_loss: 1.0116 - v

Epoch 158/200
3281/3281 [==============================] - 0s 88us/step - loss: 0.9783 - categorical_accuracy: 0.5770 - val_loss: 1.0513 - val_categorical_accuracy: 0.5152
Epoch 159/200
3281/3281 [==============================] - 0s 87us/step - loss: 0.9799 - categorical_accuracy: 0.5818 - val_loss: 1.0426 - val_categorical_accuracy: 0.5300
Epoch 160/200
3281/3281 [==============================] - 0s 78us/step - loss: 0.9887 - categorical_accuracy: 0.5724 - val_loss: 1.0666 - val_categorical_accuracy: 0.5077
Epoch 161/200
3281/3281 [==============================] - 0s 72us/step - loss: 0.9891 - categorical_accuracy: 0.5712 - val_loss: 1.0350 - val_categorical_accuracy: 0.5362
Epoch 162/200
3281/3281 [==============================] - 0s 76us/step - loss: 0.9853 - categorical_accuracy: 0.5745 - val_loss: 1.0078 - val_categorical_accuracy: 0.5436
Epoch 163/200
3281/3281 [==============================] - 0s 79us/step - loss: 0.9840 - categorical_accuracy: 0.5754 - val_loss: 1.0578 - v

Epoch 172/200
3281/3281 [==============================] - 0s 78us/step - loss: 0.9824 - categorical_accuracy: 0.5721 - val_loss: 1.0321 - val_categorical_accuracy: 0.5442
Epoch 173/200
3281/3281 [==============================] - 0s 77us/step - loss: 0.9876 - categorical_accuracy: 0.5684 - val_loss: 1.0171 - val_categorical_accuracy: 0.5337
Epoch 174/200
3281/3281 [==============================] - 0s 80us/step - loss: 0.9750 - categorical_accuracy: 0.5824 - val_loss: 1.0111 - val_categorical_accuracy: 0.5646
Epoch 175/200
3281/3281 [==============================] - 0s 73us/step - loss: 0.9744 - categorical_accuracy: 0.5800 - val_loss: 1.0437 - val_categorical_accuracy: 0.5318
Epoch 176/200
3281/3281 [==============================] - 0s 76us/step - loss: 0.9736 - categorical_accuracy: 0.5824 - val_loss: 1.0717 - val_categorical_accuracy: 0.5176
Epoch 177/200
3281/3281 [==============================] - 0s 78us/step - loss: 0.9672 - categorical_accuracy: 0.5843 - val_loss: 1.0846 - v

3281/3281 [==============================] - 0s 78us/step - loss: 0.9694 - categorical_accuracy: 0.5770 - val_loss: 1.0228 - val_categorical_accuracy: 0.5560
Epoch 187/200
3281/3281 [==============================] - 0s 81us/step - loss: 0.9668 - categorical_accuracy: 0.5837 - val_loss: 1.0386 - val_categorical_accuracy: 0.5312
Epoch 188/200
3281/3281 [==============================] - 0s 73us/step - loss: 0.9702 - categorical_accuracy: 0.5733 - val_loss: 1.0645 - val_categorical_accuracy: 0.5448
Epoch 189/200
3281/3281 [==============================] - 0s 76us/step - loss: 0.9662 - categorical_accuracy: 0.5791 - val_loss: 1.0162 - val_categorical_accuracy: 0.5498
Epoch 190/200
3281/3281 [==============================] - 0s 81us/step - loss: 0.9675 - categorical_accuracy: 0.5767 - val_loss: 1.0321 - val_categorical_accuracy: 0.5467
Epoch 191/200
3281/3281 [==============================] - 0s 79us/step - loss: 0.9667 - categorical_accuracy: 0.5840 - val_loss: 1.0489 - val_categorical

3281/3281 [==============================] - 0s 85us/step - loss: 0.9670 - categorical_accuracy: 0.5849 - val_loss: 1.0789 - val_categorical_accuracy: 0.5164
Epoch 200/200
3281/3281 [==============================] - 0s 84us/step - loss: 0.9770 - categorical_accuracy: 0.5815 - val_loss: 1.0199 - val_categorical_accuracy: 0.5325


In [94]:
preds = winemod1.evaluate(x = test_data, y = test_value)
print()
print ("Loss = " + str(preds[0]))
print ("Test Accuracy = " + str(preds[1]))

1617/1617 [==============================] - 0s 42us/step

Loss = 1.019853695204174
Test Accuracy = 0.5324675325228243


This was the result of using PCA on the input data with 10 components Accuracy = 0.3939393940315471. 

PCA did not improve the performance of the solution. This maybe due to the class imbalance problem.

If the classifier was trained on the classes separated then will get more accuracy.

This the accuracy without doing PCA Test Accuracy = 0.5324675325228243


**Linear Regression**

In [0]:
from sklearn import linear_model
lin_reg_model = linear_model.LinearRegression()

In [97]:
lin_reg_model.fit(train_data,train_value)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [0]:
predictions = lin_reg_model.predict(test_data)

In [99]:
predictions

array([[-0.04089617,  0.18397974,  0.48386163,  0.29545525,  0.08222256,
         0.00121442],
       [ 0.06346587,  0.09765265,  0.33559971,  0.39448406,  0.10804714,
         0.00262623],
       [ 0.01243089,  0.05456759,  0.50560772,  0.33783705,  0.0749814 ,
         0.00649621],
       ...,
       [ 0.03007114,  0.26865258,  0.54157182,  0.13943574,  0.02189129,
        -0.00117166],
       [ 0.04818257,  0.19727851,  0.44744237,  0.22957854,  0.0640984 ,
         0.00191709],
       [ 0.01813024,  0.33529246,  0.44269534,  0.17562613,  0.0263294 ,
        -0.00064537]])

In [102]:
classified = np.zeros_like(predictions)
classified[np.arange(len(predictions)), predictions.argmax(1)] = 1
classified

array([[0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       ...,
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0.]])

In [0]:
tvalue = np.array(test_value)

In [0]:
classified = np.array(classified)

In [0]:
from sklearn import metrics

In [0]:
score = metrics.r2_score(tvalue,classified)

In [111]:
score

-0.04525182220445245